File: backend/task-service/src/main/java/org/driventask/task/Service/TaskService.java
package org.driventask.task.Service;

import java.util.UUID;
import java.util.stream.Collectors;

import org.driventask.task.Exception.ProjectNotFoundException;
import org.driventask.task.Exception.TaskCreationException;
import org.driventask.task.Exception.TaskNotFoundException;
import org.driventask.task.Exception.TaskUpdateException;
import org.driventask.task.Exception.UserNotFoundException;
import org.driventask.task.FeignClient.ProjectClient;
import org.driventask.task.FeignClient.UserClient;
import org.driventask.task.IService.ITaskService;
import org.driventask.task.KafkaService.TaskProducer;
import org.driventask.task.Payload.Kafka.TaskCreation;
import org.driventask.task.Payload.Kafka.TaskUpdate;
import org.driventask.task.Payload.Mapper.TaskMapper;
import org.driventask.task.Payload.Request.TaskRequest;
import org.driventask.task.Payload.Response.TaskResponse;
import org.driventask.task.Repository.TaskRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
@Slf4j
@RequiredArgsConstructor
public class TaskService implements ITaskService{
    
    private final TaskMapper taskMapper;
    private final TaskRepository taskRepository;
    private final TaskProducer taskProducer;
    private final UserClient userClient;
    private final ProjectClient projectClient;

    @Override
    @Transactional
    public Mono<Void> createTask(TaskRequest taskRequest) {
        return Mono.fromCallable(() -> {
            if(!checkForProjectExistance(taskRequest.projectId())){
                return Mono.error(new ProjectNotFoundException("Project not found with ID: " + taskRequest.projectId()));
            }
            if(StringUtils.hasText(taskRequest.userAssigned()) && !checkForUserExistance(taskRequest.userAssigned())){
                return Mono.error(new UserNotFoundException("User not found with ID: " + taskRequest.userAssigned()));
            }
            return Mono.empty();
        })
        .flatMap(nullValue -> taskRepository.save(taskMapper.toTask(taskRequest)))
        .doOnNext(taskCreated -> {
            taskProducer.sendTaskCreationEvent(
                    new TaskCreation(
                        taskCreated.getId().toString(),
                        taskCreated.getTitle(),
                        taskCreated.getProjectId(),
                        taskCreated.getUserAssigned(),
                        taskCreated.getCreatedAt()
                    )
                );
        })
        .then()
        .onErrorMap(e -> new TaskCreationException("Cannot create Task"));
    }
    
    @Override
    public Mono<TaskResponse> getTaskById(String taskId) {
        return taskRepository.findById(UUID.fromString(taskId))
            .switchIfEmpty(Mono.error(new TaskNotFoundException("Cannot find Task with ID: " + taskId)))
            .map(taskMapper::fromTask);
    }

    @Override
    public Flux<TaskResponse> getAllTasks(String projectId) {
        return taskRepository.findByProjectId(projectId)
            .map(taskMapper::fromTask);
    }

    @Override
    @Transactional
    public Mono<Void> updateTask(String taskId, TaskRequest taskRequest) {
        return taskRepository.findById(UUID.fromString(taskId))
            .switchIfEmpty(Mono.error(new TaskNotFoundException("Cannot find Task")))
            .flatMap(existingTask -> {
                if(StringUtils.hasText(taskRequest.userAssigned()) && !existingTask.getUserAssigned().equals(taskRequest.userAssigned())){
                    existingTask.setUserAssigned(taskId);
                }
                existingTask.setTitle(taskRequest.title());
                existingTask.setDescription(taskRequest.description());
                existingTask.setPriority(taskRequest.priority());
                existingTask.setStatus(taskRequest.status());
                existingTask.setFilesIncluded(taskRequest.filesIncluded().stream().collect(Collectors.toList()));
                return taskRepository.save(existingTask);
            })
            .doOnNext(taskUpdated -> {
                taskProducer.sendTaskUpdateEvent(
                    new TaskUpdate(
                        taskUpdated.getId().toString(),
                        taskUpdated.getTitle(),
                        taskUpdated.getStatus(),
                        taskUpdated.getUserAssigned(),
                        taskUpdated.getProjectId(),
                        taskUpdated.getCreatedAt(),
                        taskUpdated.getUpdatedAt()
                    )
                );
            })
            .then()
            .onErrorMap(e -> new TaskUpdateException("Cannot create Task"));
    }


    @Override
    @Transactional
    public Mono<Void> deleteTask(String taskId) {
        return taskRepository.findById(UUID.fromString(taskId))
            .switchIfEmpty(Mono.error(new TaskNotFoundException("Cannot find task with id :" + taskId)))
            .flatMap(existingTask -> taskRepository.delete(existingTask))
            .then();
    }


    private boolean checkForUserExistance(String userId){
        return Boolean.valueOf(userClient.isUserExist(userId).getBody());
    }
    private boolean checkForProjectExistance(String projectId){
        return Boolean.valueOf(projectClient.isProjectExist(projectId).getBody());
    }
}

File: frontend2/src/app/services/task.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { TaskRequest } from '../models/task-request';
import { map, Observable } from 'rxjs';
import { TaskResponse } from '../models/task-response';

@Injectable({
    providedIn: 'root'
})
export class TaskService {

    private API_URL = 'http://localhost:8222/api/v1/task';
    constructor(
        private http: HttpClient
    ) { }

    createTask(task: TaskRequest): Observable<void>{
        return this.http.post<void>(this.API_URL, task);
    }

    getTaskById(id: String): Observable<TaskResponse>{
        return this.http.get<TaskResponse>(`${this.API_URL}/${id}`);
    }
    getAllTasks(projectId: String): Observable<TaskResponse[]>{
        return this.http.get<TaskResponse[]>(`${this.API_URL}/project/${projectId}`);
    }
    updateTask(id: String , taskRequest: TaskRequest): Observable<void>{
        return this.http.put<void>(`${this.API_URL}/${id}`, taskRequest);
    }
    deleteTask(id: String): Observable<void>{
        return this.http.delete<void>(`${this.API_URL}/${id}`);
    }

    /*
    createTaskWebSocket(task: TaskRequest): Observable<void> {
        return this.rxStompService.publish({ destination: '/task/create', body: JSON.stringify(task) }).pipe(
            map(() => void 0)
        );
    }

    updateTaskWebSocket(taskId: string, task: TaskRequest): Observable<void> {
        return this.rxStompService.publish({ destination: `/task/update/${taskId}`, body: JSON.stringify(task) }).pipe(
            map(() => void 0)
        );
    }

    deleteTaskWebSocket(taskId: string): Observable<void> {
        return this.rxStompService.publish({ destination: '/task/delete', body: taskId }).pipe(
            map(() => void 0)
        );
    }
        */
}